lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -12,27 +12,181 @@ namespace {
 
 //==-- main implementation goes here --==//
 class TCVR {
+  ExprMap m_cache;
+  bool m_isWellFormed = true;
+  Expr m_errorExp;
+
+  bool isConst(Expr exp) {
+    return bind::isBoolConst(exp) || bind::isIntConst(exp) ||
+           bind::isRealConst(exp);
+  }
+
+  bool isValue(Expr exp) { return isOpX<TRUE>(exp) || isOpX<FALSE>(exp); }
+
+  template <unsigned int N> bool numOfChildrenEq(Expr exp) {
+    return exp->arity() == N;
+  }
+
+  template <unsigned int N> bool numOfChildrenGreaterEq(Expr exp) {
+    return exp->arity() >= N;
+  }
+
+  bool isValidNumType(Expr exp) {
+    if (exp != nullptr &&
+        (isOp<INT_TY>(exp) || isOp<REAL_TY>(exp) || isOp<UNINT_TY>(exp)))
+      return true;
+    return false;
+  }
+
+  // ensures the expression has the correct number of children and all children
+  // are the same and correct type
+  Expr numTypeCheckChildren(Expr exp,
+                            std::function<bool(Expr exp)> checkNumChildren) {
+    if (!checkNumChildren(exp))
+      return Expr();
+
+    Expr type = m_cache.at(exp->first());
+    auto isSameType = [this, type](Expr exp) {
```

Comment:
should this by `&this`?

Reply:
It seems like you can't capture this by reference https://stackoverflow.com/questions/16323032/why-cant-i-capture-this-by-reference-this-in-lambda

Reply:
Of course. `this` is a pointer , my bad :), but `type` can be captured by reference